Refactor USD loading and enable basic animations

mwyrzykowski opened the pull request:
#### ec18d38b9142750899e62d16a8390995828023bf
<pre>
Refactor USD loading and enable basic animations
<a href="https://bugs.webkit.org/show_bug.cgi?id=303772">https://bugs.webkit.org/show_bug.cgi?id=303772</a>
<a href="https://rdar.apple.com/166085997">rdar://166085997</a>

Reviewed by Etienne Segonzac.

In order to support animations in USD models, we need to send
deformation / skinning data across IPC.

Break out some of the Swift functionality into multiple files.

Also add basic support for displaying animated USD files.

* Source/WebCore/Modules/Model/Implementation/DDMeshImpl.mm:
(WebCore::DDModel::convert):
(WebCore::DDModel::convertToList):
(WebCore::DDModel::DDMeshImpl::update):
* Source/WebCore/Modules/Model/InternalAPI/DDModel.serialization.in:
* Source/WebCore/Modules/Model/InternalAPI/DDUpdateMeshDescriptor.h:
* Source/WebCore/Modules/Model/ModelDDInlineConverters.h:
(WebCore::toCpp):
* Source/WebGPU/WebGPU/DDModelTypes.h:
* Source/WebGPU/WebGPU/USDModel.swift:
(_USDStageKit_SwiftUI.instanceTransformsCompat):
(_USDStageKit_SwiftUI.geometryBindTransformCompat):
(_USDStageKit_SwiftUI.jointTransformsCompat):
(_USDStageKit_SwiftUI.inverseBindPosesCompat):
(toWebImageAsset(_:data:)):
(convertSemantic(_:)):
(webAttributesFromAttributes(_:)):
(webLayoutsFromLayouts(_:)):
(webMeshDescriptorFromMeshDescriptor(_:)):
(webPartsFromParts(_:)):
(convert(_:)):
(webUpdateTextureRequestFromUpdateTextureRequest(_:)):
(toWebSkinningData(_:)):
(toWebBlendShapeData(_:)):
(toWebRenormalizationData(_:)):
(webDeformationDataFromData(_:)):
(webRequestInstanceTransforms):
(descriptor):
(USDModelLoader.iblTextureUpdated(_:)):
(USDModelLoader.iblTextureDestroyed(_:)):
(USDModelLoader.meshUpdated(_:)):
(USDModelLoader.materialUpdated(_:)):
(USDModelLoader.textureUpdated(_:)):
(Converter.toWebImageAsset(_:data:)): Deleted.
(Converter.convertSemantic(_:)): Deleted.
(Converter.webAttributesFromAttributes(_:)): Deleted.
(Converter.webLayoutsFromLayouts(_:)): Deleted.
(Converter.webMeshDescriptorFromMeshDescriptor(_:)): Deleted.
(Converter.webPartsFromParts(_:)): Deleted.
(Converter.convert(_:)): Deleted.
(Converter.webUpdateTextureRequestFromUpdateTextureRequest(_:)): Deleted.
(Converter.webRequestInstanceTransforms): Deleted.
(Converter.descriptor): Deleted.
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):

Canonical link: <a href="https://commits.webkit.org/305792@main">https://commits.webkit.org/305792@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a5607dfe98d09489ead7d35c2f6d26d758fa10f0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/139426 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/11802 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/928 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/147553 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/92494 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/fc9a03ef-0827-4ae1-8ca4-5f3ac564b346) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/141299 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/12509 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/11952 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/107031 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/77718 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/52fc8e3d-07f3-4756-80ea-519114764fef) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/142373 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/9555 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/124883 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/87607 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/5e4e7246-e7ca-4712-bf7b-47a9b744d80d) 
| [❌ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/138771 "Found 1 webkitpy test failure: webkit.messages_unittest.GeneratedFileContentsTest.test_message_argument_description") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/9162 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/6808 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/7850 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/118494 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/167/builds/851 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/150336 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/11486 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/867 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/115145 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/11499 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/9801 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/115456 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/29328 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/9893 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/121297 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/66492 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/11529 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/813 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/11264 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/75196 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/11466 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/11316 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WebGPU/USDModel.swift:
```diff
@@ -32,6 +32,27 @@ import WebGPU_Internal
 import _USDStageKit_SwiftUI
 import ShaderGraph
+
+extension _USDStageKit_SwiftUI._Proto_MeshDataUpdate_v1 {
```
@rr-codes I can leave these changes out of this patch, or fold them into https://github.com/WebKit/WebKit/pull/54993/files, since modifying the SDK swift interface file makes them not strictly necessary. 

But it appears to be a Swift compiler issue

mwyrzykowski commented on Source/WebGPU/WebGPU/USDModel.swift:
```diff
@@ -1049,146 +1154,209 @@ extension DDBridgeReceiver {
     }
 }
 
-final class Converter {
```
'ignore whitespace' setting in Githubs helps reduce the change markers, since I just moved these to freestanding functions

rr-codes commented on Source/WebGPU/WebGPU/USDModel.swift:
```diff
@@ -1049,146 +1154,209 @@ extension DDBridgeReceiver {
-            arrayLength: asset.arrayLength,
-            textureUsage: asset.textureUsage,
-            swizzle: asset.swizzle
+nonisolated func toWebImageAsset(_ asset: LowLevelTexture.Descriptor, data: Data) -> DDBridgeImageAsset {
```
please make these extension initializers on DDBridgeImageAsset

rr-codes commented on Source/WebGPU/WebGPU/USDModel.swift:
```diff
@@ -1049,146 +1154,209 @@ extension DDBridgeReceiver {
+    }
+}
+
+nonisolated func webAttributesFromAttributes(_ attributes: [LowLevelMesh.Attribute]) -> [DDBridgeVertexAttributeFormat] {
```
why non isolated?

mwyrzykowski replied:
oh good question, its because I didn't realize I could just make it `fileprivate` - will remove the `nonisolated`

rr-codes commented on Source/WebGPU/WebGPU/USDModel.swift:
```diff
@@ -1049,146 +1154,209 @@ extension DDBridgeReceiver {
+
+nonisolated func convertSemantic(_ semantic: LowLevelMesh.VertexSemantic) -> Int {
+    switch semantic {
+    case .position: return 0
```
this is existing code, but you can omit the returns

rr-codes commented on Source/WebGPU/WebGPU/USDModel.swift:
```diff
@@ -1049,146 +1154,209 @@ extension DDBridgeReceiver {
+    )
+}
+
+nonisolated func toData<T>(_ input: [T]) -> Data {
```
instead of this function, does `Data(buffer: &whateverArray)` work?

mwyrzykowski replied:
Swift does not seem to like that:
```
/Volumes/Data/OpenSource/Source/WebGPU/WebGPU/USDModel.swift:1316:5: error: no exact matches in call to initializer 
    Data(buffer: &input)
/Volumes/Data/OpenSource/Source/WebGPU/WebGPU/USDModel.swift:1316:5: note: found candidate with type '(UnsafeBufferPointer<SourceType>) -> Data'
    Data(buffer: &input)
/Volumes/Data/OpenSource/Source/WebGPU/WebGPU/USDModel.swift:1316:5: note: found candidate with type '(UnsafeMutableBufferPointer<SourceType>) -> Data'
    Data(buffer: &input)
/Volumes/Data/OpenSource/Source/WebGPU/WebGPU/USDModel.swift:1316:18: error: cannot pass immutable value as inout argument: 'input' is a 'let' constant
    Data(buffer: &input)
```

rr-codes commented on Source/WebGPU/WebGPU/USDModel.swift:
```diff
@@ -1049,146 +1154,209 @@ extension DDBridgeReceiver {
+    input.map { toData($0) }
+}
+
+nonisolated func toWebSkinningData(_ request: _Proto_DeformationData_v1.SkinningData?) -> DDBridgeSkinningData? {
```
please make this an extension initializer on DDBridgeSkinningData

rr-codes commented on Source/WebGPU/WebGPU/USDModel.swift:
```diff
@@ -1049,146 +1154,209 @@ extension DDBridgeReceiver {
-            indexCapacity: request.indexCapacity,
-            indexType: request.indexType
-        )
+nonisolated func toWebRenormalizationData(_ request: _Proto_DeformationData_v1.RenormalizationData?) -> DDBridgeRenormalizationData? {
```
ditto

rr-codes commented on Source/WebGPU/WebGPU/USDModel.swift:
```diff
@@ -1049,146 +1154,209 @@ extension DDBridgeReceiver {
-                boundsMax: a.bounds.max
-            )
-        })
+nonisolated func webDeformationDataFromData(_ request: _Proto_DeformationData_v1?) -> DDBridgeDeformationData? {
```
ditto

rr-codes commented on Source/WebGPU/WebGPU/USDModel.swift:
```diff
@@ -32,6 +32,27 @@ import WebGPU_Internal
+
+extension _USDStageKit_SwiftUI._Proto_MeshDataUpdate_v1 {
+    @_silgen_name("$s20_USDStageKit_SwiftUI24_Proto_MeshDataUpdate_v1V18instanceTransformsSaySo13simd_float4x4aGvg")
+    fileprivate func instanceTransformsCompat() -> [simd_float4x4]
```
does this have to be a function? or can it be a property like the actual real property its simulating?

mwyrzykowski replied:
I had tried that initially, but Swift says:
```
Extensions must not contain stored properties
```

rr-codes commented on Source/WebGPU/WebGPU/USDModel.swift:
```diff
@@ -1049,146 +1154,209 @@ extension DDBridgeReceiver {
-    ) -> DDBridgeUpdateMaterial {
-        DDBridgeUpdateMaterial(materialGraph: request.materialSourceArchive, identifier: request.identifier)
+    func iblTextureDestroyed(identifier: _Proto_ResourceId) {
+        //
```
is this comment intentional?

mwyrzykowski replied:
yes but there should be a fixme too, I will add one

rr-codes commented on Source/WebGPU/WebGPU/USDModel.swift:
```diff
@@ -1237,7 +1405,7 @@ final class USDModelLoader: _Proto_UsdStageSession_v1.Delegate {
         let identifier = data.identifier
         self.dispatchSerialQueue.async {
-            self.objcLoader.updateTexture(webRequest: Converter.webUpdateTextureRequestFromUpdateTextureRequest(data))
+            self.objcLoader.updateTexture(webRequest: webUpdateTextureRequestFromUpdateTextureRequest(data))
```
this file should be split up probably, it’s starting to get pretty big and multiple responsibilities

mwyrzykowski replied:
that is a good idea, perhaps starting with the bridging types?

I filed https://bugs.webkit.org/show_bug.cgi?id=303907 to track that

mwyrzykowski replied:
I broke up the files as part of this PR

rr-codes commented on Source/WebGPU/WebGPU/USDModel.swift:
```diff
@@ -140,6 +161,87 @@ extension DDBridgeChainedFloat4x4 {
+@objc
+@implementation
+extension DDBridgeSkinningData {
+    let influencePerVertexCount: UInt8
```
why do all of these actually need to be Data/bytes?

mwyrzykowski replied:
Swift has no way I'm aware of to pass an array of primitive types to Objective-C

etiennesegonzac commented on Source/WebGPU/WebGPU/DDMesh.h:
```diff
@@ -84,6 +84,7 @@ class DDMesh : public ThreadSafeRefCounted<DDMesh>, public WGPUDDMeshImpl {
 #endif
     NSUUID* m_ddMeshIdentifier;
     mutable uint32_t m_currentTexture { 0 };
+    bool meshDataExists { false };
```
this should be in the `#if ENABLE(GPU_PROCESS_MODEL)` block

etiennesegonzac commented on Source/WebCore/Modules/model-element/DDModelPlayer.mm:
```diff
@@ -157,7 +170,34 @@ void setOpaque(bool opaque)
+        .swizzle = { }
+    };
+    DDModel::DDImageAsset specularTexture {
+        .data = loadData(adoptCF((CFStringRef)@"modelSpecularData")),
```
should we name the files `modelDefault*Data`?
(assuming they won't be used when using custom IBL)

etiennesegonzac commented on Source/WebGPU/WebGPU/ModelUtils.swift:
```diff
@@ -0,0 +1,127 @@
+@_spi(UsdLoaderAPI) import _USDStageKit_SwiftUI
+
+nonisolated func mapSemantic(_ semantic: LowLevelMesh.VertexSemantic) -> _Proto_LowLevelMeshResource_v1.VertexSemantic {
+    switch semantic {
```
single satement, no need for the returns

etiennesegonzac commented on Source/WebGPU/WebGPU/USDModel.swift:
```diff
@@ -465,371 +143,528 @@ nonisolated func makeTextureFromImageAsset(
-    }
-}
+        guard let inputMeshDescription else {
+            logError("inputMeshDescription is unexepectedly nil")
```
unexpectedly?
(got copy/pasted a few times)

etiennesegonzac reviewed (commented):
What's the latest?
Do we plan on splitting out the bridging types (or the default IBL introduction?) from the deformation data or should I review the mega-patch?
(In which case we should probably re-title.)

etiennesegonzac commented on Source/WebGPU/WebGPU/ModelRenderer.swift:
I think some (older) versions of the compiler don't like the trailing commas

etiennesegonzac commented on Source/WebGPU/WebGPU/ModelRenderer.swift:
ditto

etiennesegonzac commented on Source/WebGPU/WebGPU/USDModel.swift:
should we keep the gpu capture out of the build and make it easier to toggle locally?

etiennesegonzac commented on Source/WebGPU/WebGPU/USDModel.swift:
fail to render?

etiennesegonzac commented on Source/WebGPU/WebGPU/USDModel.swift:
dupe from L776?

etiennesegonzac commented on Source/WebGPU/WebGPU/USDModel.swift:
unused?

etiennesegonzac commented on Source/WebGPU/WebGPU/USDModel.swift:
unused?

etiennesegonzac commented on Source/WebKit/WebProcess/GPU/graphics/Model/RemoteDDMeshProxy.cpp:
This feels too dependent on `computeMinAndMaxCorners` internals / initialization. Can we make it more obvious? (or at least add a comment :))

etiennesegonzac commented on Source/WebGPU/WebGPU/ModelBridge.swift:
This extension is a bit boilerplate-y, could we extract a generic helper like
`private func extractArray<T>(from data: Data?, debugName: String) -> [T]`

mwyrzykowski replied:
oh this is a good idea, I had no idea that was possible in Swift

etiennesegonzac commented on Source/WebGPU/WebGPU/USDModel.swift:
nit: should this really be a logError?

etiennesegonzac commented on Source/WebCore/Modules/model-element/DDModelPlayer.mm:
just double-checking this is intentional :)

mwyrzykowski replied:
Yes

mwyrzykowski replied:
It allows us to specify transparent background colors

etiennesegonzac reviewed (approved):
There's a typo in the commit message, "Rector USD" :)
Some surface-level comments, but I think this is ready to go!

rr-codes commented on Source/WebGPU/WebGPU/ModelRenderer.swift:
what’s up with this?

mwyrzykowski replied:
More Swift compiler bugs it seems. It can’t find an extension method when another one with fewer parameters exist

rr-codes commented on Source/WebGPU/WebGPU/ModelRenderer.swift:
internal is not needed; it’s the default

rr-codes commented on Source/WebGPU/WebGPU/ModelRenderer.swift:
```suggestion
final class Renderer {
```

rr-codes commented on Source/WebGPU/WebGPU/ModelRenderer.swift:
no need for internal

rr-codes commented on Source/WebGPU/WebGPU/ModelUtils.swift:
these should all be convenience initializers

rr-codes commented on Source/WebGPU/WebGPU/ModelUtils.swift:
```suggestion
        if let indexData {
```

rr-codes commented on Source/WebGPU/WebGPU/ModelUtils.swift:
please use guard let … else { fatalError(“explanation”)

mwyrzykowski committed:
Refactor USD loading and enable basic animations
https://bugs.webkit.org/show_bug.cgi?id=303772
rdar://166085997

Reviewed by Etienne Segonzac.

In order to support animations in USD models, we need to send
deformation / skinning data across IPC.

Break out some of the Swift functionality into multiple files.

Also add basic support for displaying animated USD files.

* Source/WebCore/Modules/Model/Implementation/DDMeshImpl.mm:
(WebCore::DDModel::convert):
(WebCore::DDModel::convertToList):
(WebCore::DDModel::DDMeshImpl::update):
* Source/WebCore/Modules/Model/InternalAPI/DDModel.serialization.in:
* Source/WebCore/Modules/Model/InternalAPI/DDUpdateMeshDescriptor.h:
* Source/WebCore/Modules/Model/ModelDDInlineConverters.h:
(WebCore::toCpp):
* Source/WebGPU/WebGPU/DDModelTypes.h:
* Source/WebGPU/WebGPU/USDModel.swift:
(_USDStageKit_SwiftUI.instanceTransformsCompat):
(_USDStageKit_SwiftUI.geometryBindTransformCompat):
(_USDStageKit_SwiftUI.jointTransformsCompat):
(_USDStageKit_SwiftUI.inverseBindPosesCompat):
(toWebImageAsset(_:data:)):
(convertSemantic(_:)):
(webAttributesFromAttributes(_:)):
(webLayoutsFromLayouts(_:)):
(webMeshDescriptorFromMeshDescriptor(_:)):
(webPartsFromParts(_:)):
(convert(_:)):
(webUpdateTextureRequestFromUpdateTextureRequest(_:)):
(toWebSkinningData(_:)):
(toWebBlendShapeData(_:)):
(toWebRenormalizationData(_:)):
(webDeformationDataFromData(_:)):
(webRequestInstanceTransforms):
(descriptor):
(USDModelLoader.iblTextureUpdated(_:)):
(USDModelLoader.iblTextureDestroyed(_:)):
(USDModelLoader.meshUpdated(_:)):
(USDModelLoader.materialUpdated(_:)):
(USDModelLoader.textureUpdated(_:)):
(Converter.toWebImageAsset(_:data:)): Deleted.
(Converter.convertSemantic(_:)): Deleted.
(Converter.webAttributesFromAttributes(_:)): Deleted.
(Converter.webLayoutsFromLayouts(_:)): Deleted.
(Converter.webMeshDescriptorFromMeshDescriptor(_:)): Deleted.
(Converter.webPartsFromParts(_:)): Deleted.
(Converter.convert(_:)): Deleted.
(Converter.webUpdateTextureRequestFromUpdateTextureRequest(_:)): Deleted.
(Converter.webRequestInstanceTransforms): Deleted.
(Converter.descriptor): Deleted.
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):

Canonical link: https://commits.webkit.org/305792@main
